librocksdb-sys/patches/rocksdb.cc

```diff
@@ -0,0 +1,111 @@
+#include "patches/rocksdb.h"
+
+#include "rocksdb/utilities/options_util.h"
+
+using rocksdb::Cache;
+using rocksdb::ColumnFamilyDescriptor;
+using rocksdb::Env;
+using rocksdb::Options;
+using rocksdb::Status;
+
+extern "C" {
+    // Copy structs from librocksdb-sys/rocksdb/db/c.cc
+    struct rocksdb_cache_t {
+        std::shared_ptr<Cache>  rep;
+    };
+    struct rocksdb_env_t {
+        Env* rep;
+        bool is_default;
+    };
+    struct rocksdb_options_t {
+        Options rep;
+    };
+
+    // New structs
+    struct rocksdb_column_family_descriptor_t {
+        char *name;
+        Options options;
+    };
+    struct rocksdb_column_family_descriptors_t {
+        std::vector<rocksdb_column_family_descriptor_t> rep;
+    };
+
+    rocksdb_cache_t* rocksdb_null_cache() {
+        rocksdb_cache_t* c = new rocksdb_cache_t;
+        c->rep = nullptr;
+        return c;
+    }
+
+    rocksdb_options_t* rocksdb_options_clone(rocksdb_options_t* options) {
+        rocksdb_options_t* o = new rocksdb_options_t;
+        o->rep = Options(options->rep);
+        return o;
+    }
+
+    rocksdb_column_family_descriptors_t* rocksdb_column_family_descriptors_create() {
+        return new rocksdb_column_family_descriptors_t;
+    }
+
+    void rocksdb_column_family_descriptors_destroy(rocksdb_column_family_descriptors_t* cf_descs) {
+        int size = static_cast<int>(cf_descs->rep.size());
+        for (int i = 0; i < size; i++) {
+            free(cf_descs->rep[i].name);
+        }
+        delete cf_descs;
+    }
+
+    int rocksdb_column_family_descriptors_count(const rocksdb_column_family_descriptors_t* cf_descs) {
+        return static_cast<int>(cf_descs->rep.size());
+    }
+
+    char* rocksdb_column_family_descriptors_name(const rocksdb_column_family_descriptors_t* cf_descs, int index) {
+        return cf_descs->rep[index].name;
+    }
+
+    rocksdb_options_t* rocksdb_column_family_descriptors_options(const rocksdb_column_family_descriptors_t* cf_descs, int index) {
+        rocksdb_options_t* options = new rocksdb_options_t;
+        options->rep = cf_descs->rep[index].options;
+        return options;
+    }
+
+    rocksdb_fulloptions_t rocksdb_options_load_from_file(
+        const char* config_file,
+        rocksdb_env_t* env,
+        bool ignore_unknown_options,
+        rocksdb_cache_t* cache,
+        char** errptr) {
+
+        rocksdb_fulloptions_t full_opts;
+        full_opts.db_opts = nullptr;
+        full_opts.cf_descs = nullptr;
+
+        rocksdb_options_t* db_opts = new rocksdb_options_t;
+        std::vector<ColumnFamilyDescriptor> cf_descs_tmp;
+
+        Status status = rocksdb::LoadOptionsFromFile(
+            std::string(config_file),
+            env->rep,
+            &db_opts->rep,
+            &cf_descs_tmp,
+            ignore_unknown_options,
+            &cache->rep);
+        if (status.ok()) {
```

Comment:
`db_opts` is leaked when status is not OK.

Reply:
Please use some tools to check the memory issues

Reply:
### Brief

- I fix this issue.
  https://github.com/nervosnetwork/rust-rocksdb/blob/4c594ea77b3647d6dac6900e6c4575f1bfd2f231/librocksdb-sys/patches/rocksdb.cc#L104-L106
- I add a target `test` in `Makefile` to detect memory leaks for this function which I added in this PR.

### Report

<details><summary>Chick this to read the memory leak report.(run <code>cd librocksdb-sys && make test</code>)</summary>
<pre>
rm -rf tmp-test-memory-leak
gcc -g -I rocksdb/include/ -I . \
                -L $(find . -name "librocksdb.a" -exec dirname {} \; | head -1) \
                -o tmp-test-memory-leak \
                tests/memory/rocksdb_options_load_from_file.c \
                -lrocksdb -lpthread -ldl -lm -lstdc++ \
        && valgrind --tool=memcheck --leak-check=full ./tmp-test-memory-leak \
    && rm tmp-test-memory-leak
==18507== Memcheck, a memory error detector
==18507== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==18507== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==18507== Command: ./tmp-test-memory-leak
==18507== 
==18507== 
==18507== HEAP SUMMARY:
==18507==     in use at exit: 12,422 bytes in 50 blocks
==18507==   total heap usage: 3,843 allocs, 3,793 frees, 378,726 bytes allocated
==18507== 
==18507== 16 bytes in 1 blocks are definitely lost in loss record 5 of 29
==18507==    at 0x4836E0F: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==18507==    by 0x151E06: rocksdb_create_default_env (c.cc:3339)
==18507==    by 0x142388: main (rocksdb_options_load_from_file.c:8)
==18507== 
==18507== 528 (16 direct, 512 indirect) bytes in 1 blocks are definitely lost in loss record 24 of 29
==18507==    at 0x4836E0F: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==18507==    by 0x151B45: rocksdb_cache_create_lru (c.cc:3306)
==18507==    by 0x14239A: main (rocksdb_options_load_from_file.c:10)
==18507== 
==18507== 3,602 (24 direct, 3,578 indirect) bytes in 1 blocks are definitely lost in loss record 27 of 29
==18507==    at 0x4836E0F: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==18507==    by 0x4DA85E: rocksdb_options_load_from_file (rocksdb.cc:93)
==18507==    by 0x1423BD: main (rocksdb_options_load_from_file.c:13)
==18507== 
==18507== 8,044 (1,168 direct, 6,876 indirect) bytes in 1 blocks are definitely lost in loss record 29 of 29
==18507==    at 0x4836E0F: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==18507==    by 0x4DA782: rocksdb_options_load_from_file (rocksdb.cc:82)
==18507==    by 0x1423BD: main (rocksdb_options_load_from_file.c:13)
==18507== 
==18507== LEAK SUMMARY:
==18507==    definitely lost: 1,224 bytes in 4 blocks
==18507==    indirectly lost: 10,966 bytes in 45 blocks
==18507==      possibly lost: 0 bytes in 0 blocks
==18507==    still reachable: 232 bytes in 1 blocks
==18507==         suppressed: 0 bytes in 0 blocks
==18507== Reachable blocks (those to which a pointer was found) are not shown.
==18507== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==18507== 
==18507== For lists of detected and suppressed errors, rerun with: -s
==18507== ERROR SUMMARY: 4 errors from 4 contexts (suppressed: 0 from 0)
</pre>
</details>

### Explanations

- The 1st memory leak is caused by an environment, it will free in Rust.
  https://github.com/nervosnetwork/rust-rocksdb/blob/65f16a7cd7591d72526ed72d3413c348fe6bab95/src/options.rs#L45-L52
- The 2nd memory leak is caused by a cache. After call this function, we will lost control of it, so we can't free it.
- The 3rd memory leak is caused by column family descriptors, they will free in Rust.
  https://github.com/nervosnetwork/rust-rocksdb/blob/65f16a7cd7591d72526ed72d3413c348fe6bab95/src/options.rs#L67
  https://github.com/nervosnetwork/rust-rocksdb/blob/65f16a7cd7591d72526ed72d3413c348fe6bab95/librocksdb-sys/patches/rocksdb.cc#L43-L49
- The 4th memory leak is caused by an options, it will free in Rust.
  https://github.com/nervosnetwork/rust-rocksdb/blob/65f16a7cd7591d72526ed72d3413c348fe6bab95/src/db_options.rs#L233-L239